src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -78,30 +80,144 @@ import java.util.*
  *      ```
  */
 
-
-
-
 private val currentTime: LocalDateTime get() = LocalDateTime.now()
 
 private fun LocalDateTime.prettyPrint(): String {
     val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")
     return this.format(formatter)
 }
 
+enum class TransactionType {
+    DEPOSIT, WITHDRAWAL
+}
+
+enum class TransactionStatus {
+    SUCCESS, FAILURE
+}
+
+data class Transaction(
+    val transactionDate: LocalDateTime,
+    val transactionType: TransactionType,
+    val amount: Double,
+    val oldBalance: Double,
+    val newBalance: Double,
+    val transactionStatus: TransactionStatus
+)
+
+class TransactionalBankAccount(
+    private val accountNumber: String,
+    private val accountHolderName: String,
+    private var balance: Double = 0.0
+) : BankAccount(accountNumber, accountHolderName, balance) {
+
+    private var transactions = mutableListOf<Transaction>()
+
+    override fun deposit(amount: Double) {
+        transactions.add(
+            Transaction(
+                currentTime,
+                TransactionType.DEPOSIT,
+                amount,
+                getBalance(),
+                getBalance() + amount,
+                TransactionStatus.SUCCESS
+            )
+        )
+        super.deposit(amount)
+    }
+
+    override fun withdraw(amount: Double): Boolean {
+        val oldBalance = getBalance()
+        val success = super.withdraw(amount)
+
+        transactions.add(
+            Transaction(
+                transactionDate = currentTime,
+                transactionType = TransactionType.WITHDRAWAL,
+                amount = amount,
+                oldBalance = oldBalance,
+                newBalance = getBalance(),
+                transactionStatus = if (success) TransactionStatus.SUCCESS else TransactionStatus.FAILURE,
+            )
+        )
+        return success;
+    }
+
+    fun getAllTranactions(): List<Transaction> {
+        return transactions;
```

Comment:
There are two problems with this: 
1. The returned list is not sorted descending by date
2. We break encapsulation of the class and leak it's internal state, because we return the same list reference
Here is example that explains how encapsulation is broken and how internal state is leaked:
```
fun main() {
    val a = A()
    val transactions = a.transactions() as MutableList<Int>
    println(a.transactions()) // prints [1]
    transactions.add(2)
    println(a.transactions()) // prints [1, 2] even though we modified list externally.
}

class A {
    private val transactions = mutableListOf<Int>(1)
    
    fun transactions(): List<Int> = transactions
}
```